source/extensions/transport_sockets/tls/utility.cc

```diff
@@ -71,13 +71,60 @@ Envoy::Ssl::CertificateDetailsPtr Utility::certificateDetails(X509* cert, const
   return certificate_details;
 }
 
+bool Utility::labelWildcardMatch(absl::string_view dns_label, absl::string_view pattern) {
+  if (pattern.empty()) {
+    return dns_label.empty();
+  }
+
+  // Only valid if wildcard character appear once
+  if (std::count(pattern.begin(), pattern.end(), '*') == 1) {
+    constexpr char glob = '*';
+    // Check the special case of a single * pattern, as it's common.
+    if (pattern.size() == 1 && pattern[0] == glob) {
+      return true;
+    }
+    size_t i = 0, p = 0, i_star = dns_label.size(), p_star = 0;
+    while (i < dns_label.size()) {
+      if (p < pattern.size() &&
+          absl::ascii_tolower(dns_label[i]) == absl::ascii_tolower(pattern[p])) {
+        ++i;
+        ++p;
+      } else if (p < pattern.size() && pattern[p] == glob) {
+        i_star = i;
+        p_star = p++;
+      } else if (i_star != dns_label.size()) {
+        i = ++i_star;
+        p = p_star + 1;
+      } else {
+        return false;
+      }
+    }
+
+    while (p < pattern.size() && pattern[p] == glob) {
+      ++p;
+    }
+
+    return p == pattern.size() && i == dns_label.size();
+  }
+
+  return false;
+}
+
 bool Utility::dnsNameMatch(absl::string_view dns_name, absl::string_view pattern) {
+  // A-label ACE prefix https://www.rfc-editor.org/rfc/rfc5890#section-2.3.2.5
+  constexpr absl::string_view ACE_prefix = "xn--";
   const std::string lower_case_dns_name = absl::AsciiStrToLower(dns_name);
   const std::string lower_case_pattern = absl::AsciiStrToLower(pattern);
   if (lower_case_dns_name == lower_case_pattern) {
     return true;
   }
 
+  // https://www.rfc-editor.org/rfc/rfc6125#section-6.4.3 part #2
+  // If the wildcard character is the only character of the left-most
+  // label in the presented identifier, the client SHOULD NOT compare
+  // against anything but the left-most label of the reference identifier
+  // (e.g., *.example.com would match foo.example.com but
+  //       not bar.foo.example.com or example.com)
   size_t pattern_len = lower_case_pattern.length();
```

Comment:
Is this case needed anymore? Isn't this a subset of the new wildcard matching?

Reply:
That is true that the new wildcard matching will cover whatever is done here as well. I tested by removing from L128 to L136. The unit test passed for all the cases.

I just kept it so that for test case with `*.example.com` patter we would get the results faster. But if that is anti pattern then I can go ahead and remove these cases. Also probably simplify the code comments (should be apparent for the reader from just the rfc link).

Reply:
Removed this case, as it is not needed anymore.

I will wait for the CI to complete.